### DISCUSSION TITLE
Template Repos and Commit History

### DISCUSSION THREAD

**Discussion by steven-rothwell** [community/community#50012] [Repositories] [answered]
### Select Topic Area

Bug

### Body

When I create a new repo from a template repo, the new repo's history starts at the point of creation, it does not include all of the template repo's history.  But, when I make an update to the template repo and merge the changes into the new repo, it is including all of the template repo's history.  Why is it not just taking the commits from the point of the new repo's creation forward?

How to I produced this issue:
- Created a template repo named StandardApi.
- Made multiple commits into StandardApi's main so that there are multiple commits in the history.
- Created a new repo from StandardApi called NewApi.
- I made one commit into NewApi. (I added a branch new controller NewController.cs.)
- In StandardApi, I updated .NET from 6 to 7, which is an extremely simple change. Merged that into main.
- In NewApi, I created a feature branch (updatesFromTemplate) to pull in the template's updates. Then ran:
> - git remote add template [url to StandardApi]
> - git fetch --all
> - git merge template/main --allow-unrelated-histories
- Created a PR for updatesFromTemplate into NewApi's main. Noticed two oddities:
> - Had a merge conflict in the .csproj stating both were changed, but in reality this should be a forward commit with no conflict.
> - All of the commit history from StandardApi is being included in the merge.
- The net result seems correct in that the only changed file in the PR is the csproj change from 6 to 7. But, the above oddities make it seem like it is not working 100% as expected.

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @steven-rothwell ! 

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Issues category is solely related to conversations around the GitHub products Issues and Projects. I've gone ahead and moved it for you. Good luck!

**Comment by airtower-luna** [comment, marked as answer]
> When I create a new repo from a template repo, the new repo's history starts at the point of creation, it does not include all of the template repo's history.

Technically that is the answer already. The new repository is a *completely new repository* to Git. There's no shared history, so when you merge against the warning you get *all* commits from the repository you merge with. Generally when you'd need `--allow-unrelated-histories` to merge you should stop and ask if your intention is to combine projects that used to be completely unrelated (maybe because you're fundamentally restructuring your projects). If that's not the case something likely went wrong way before the merge.

If you care about being able to merge from the upstream repository in a way that makes sense creating a new repository from the template is probably the wrong tool, and it'd be better to start from a full copy (or fork) of the upstream repository. That way you'll be able to merge as usual. Whether it's better to use a fork or create an empty repository and push from a local clone of the upstream repo depends on if/how you want to interact with upstream.

**Reply by steven-rothwell** [reply]
Thank you for that explanation.  I am just learning about template repos and thought they worked differently than intended.
